projects/feature-activation/0004-phased-testing.md

```diff
@@ -0,0 +1,206 @@
+        return
+
+    feature_descriptions = self.feature_service.get_bits_description(block=vertex)
+    state_by_feature = {
+        feature.value: description.state.value
+        for feature, description in feature_descriptions.items()
+    }
+
+    self.log.info(
+        'New block accepted with feature activation states',
+        block_height=vertex.get_height(),
+        features_states=state_by_feature
+    )
+```
+
+This logging will be removed after Phase 1 is complete.
+
+## Phase 2
+
+Phase 2 can start as soon as Phase 1 ends. Let `M` be a block height for the beginning of Phase 2. It'll define a new `INCREASE_MAX_MERKLE_PATH_LENGTH` feature that will actually affect how the full node verifies merged mining blocks.
+
+Currently, there's a `MAX_MERKLE_PATH_LENGTH: int = 12` constant in `aux_pow.py`. This constant will be replaced by a function call to `get_max_merkle_path_length()`, with its reference implementation as follows:
+
+```python
+def get_max_merkle_path_length(feature_service: FeatureService, block: Block) -> int:
+    if feature_service.is_feature_active(
+      block=block,
+      feature=Feature.INCREASE_MAX_MERKLE_PATH_LENGTH
+    ):
+        return NEW_MAX_MERKLE_PATH_LENGTH
+
+    return OLD_MAX_MERKLE_PATH_LENGTH
+```
+
+Where this function is actually defined will be left as an implementation detail. TODO: what's a good value for `NEW_MAX_MERKLE_PATH_LENGTH`?
+
+Then, its reference configuration is:
+
+```yml
+# testnet.yml
+
+FEATURE_ACTIVATION:
+  default_threshold: 30240 # 30240 = 75% of evaluation_interval (40320)
+  features:
+    INCREASE_MAX_MERKLE_PATH_LENGTH:
+      bit: 3
+      start_height: <M>
+      timeout_height: <M + 2 * 40320>
+      minimum_activation_height: 0
+      lock_in_on_timeout: false
+      version: <insert the hathor-core version>
+      signal_support_by_default: true
+```
+
+And it's expected to reach its threshold and become active in 4 weeks (2 in `STARTED`, 2 in `LOCKED_IN`). We'll use the same validation mechanisms described in Phase 1 to make sure the feature states transition as expected. Also, we should configure a merged miner to mine blocks with a higher merkle path length, making sure its blocks are invalidated before the feature becomes `ACTIVE`, and validated after it does.
+
+## Phase 3
+
+Phase 3 can start as soon as Phase 1 ends, and it is completely analogous to Phase 1, except the NOP features will be configured on `mainnet.yml` instead of `testnet.yml`, and that `N` will be different.
+
+## Conclusion
+
+As described above, Phase 1 takes 6 weeks, and Phase 2 and Phase 3 can run concurrently, respectively for 6 and 4 weeks. This makes for a total Phased Testing duration of 12 weeks. If this duration is deemed too high, we can actually shift Phase 2 and 3 earlier, as they do not conflict with Phase 1 in any way. Another option to make it even faster, is decreasing the `EVALUATION_INTERVAL`, for example from 2 weeks to 1 week, only in testnet.
```

Comment:
Should we test a feature deactivation after a rebase on testnet only? As the hash rate is so small, we should be able to do it easily. Or we can create a private network just to get this tested.

Reply:
Did you mean reorg? If yes, then I agree it makes sense and I'll add it

Reply:
Yes, reorg. My bad.

Reply:
Thanks!